### ISSUE TITLE
Why adding ground truth bbbox during training?

### ISSUE THREAD

**Issue by MaureenZOU** [facebookresearch/maskrcnn-benchmark#570]
## ❓ Questions and Help
Could anyone illustrate what is the intuition of add the ground truth bbox during training?
        
# append ground-truth bboxes to proposals
        if self.training and targets is not None:
            boxlists = self.add_gt_proposals(boxlists, targets)

Thanks a lot!

**Comment by LeviViana** [comment]
I believe it is a hack to provide data to `roi_heads` classification task during training. These ground-truth bboxes are not taken into account in the rpn loss, as you can see [here](https://github.com/facebookresearch/maskrcnn-benchmark/blob/6b1ab0179425169158b1830cbad638f395a7defd/maskrcnn_benchmark/modeling/rpn/rpn.py#L114-L120), buy they are taken into account in the `roi_heads` loss, as you can see [here](https://github.com/facebookresearch/maskrcnn-benchmark/blob/6b1ab0179425169158b1830cbad638f395a7defd/maskrcnn_benchmark/modeling/detector/generalized_rcnn.py#L50-L52).

Once they are fed into the `roi_heads` loss evaluator, they will only generator meaningful information to the [classification loss](https://github.com/facebookresearch/maskrcnn-benchmark/blob/6b1ab0179425169158b1830cbad638f395a7defd/maskrcnn_benchmark/modeling/roi_heads/box_head/loss.py#L146), since its bounding boxes must match perfectly to the targets, no regression loss is generated, then no gradients either.

Thus, this feeds your `roi_heads` with some useful information to train even in the early stages of the training process.

I'm not sure though this is the only reason.

**Comment by MaureenZOU** [comment]
Thanks a lot, that explanation makes sense to me!

**Comment by chengyangfu** [comment]
This definitely can make the model converge faster. Otherwise,  the training of `roi_heads` would wait until RPN converge first.  Besides,  it can also make sure the input to the `roi_heads` is not empty. If the input is empty, it will cause some problems now and hard to solve it now.

**Comment by LeviViana** [comment]
I'm closing this issue, please let me know if you have any further question.

**Comment by abcxs** [comment]
> This definitely can make the model converge faster. Otherwise, the training of `roi_heads` would wait until RPN converge first. Besides, it can also make sure the input to the `roi_heads` is not empty. If the input is empty, it will cause some problems now and hard to solve it now.
I don't think it will cause the input to roi_heads become empty, rpn will select some proposals in any case
